File: src/main/java/org/voegtle/weatherwidget/widget/SmallWidgetScreenPainter.java
package org.voegtle.weatherwidget.widget;

import android.appwidget.AppWidgetManager;
import android.widget.RemoteViews;
import org.voegtle.weatherwidget.R;
import org.voegtle.weatherwidget.data.WeatherData;
import org.voegtle.weatherwidget.preferences.ColorScheme;
import org.voegtle.weatherwidget.util.ColorUtil;

public class SmallWidgetScreenPainter extends AbstractWidgetScreenPainter {
  private RemoteViews remoteViews;

  public SmallWidgetScreenPainter(AppWidgetManager appWidgetManager, int[] widgetIds, RemoteViews remoteViews) {
    super(appWidgetManager, widgetIds, remoteViews);
    this.remoteViews = remoteViews;
  }

  @Override
  public void showDataIsInvalid() {
    remoteViews.setTextColor(R.id.weather_small, ColorUtil.INSTANCE.updateColor(ColorScheme.dark));
  }

  @Override
  public void showDataIsValid() {
    updateAllWidgets();
  }

  public void updateData(WeatherData data) {
    if (data != null) {
      remoteViews.setTextViewText(R.id.weather_small, formatter.formatTemperature(data));
      remoteViews.setTextColor(R.id.weather_small, ColorUtil.INSTANCE.byAge(data.getTimestamp()));
    } else {
      remoteViews.setTextColor(R.id.weather_small, ColorUtil.INSTANCE.outdatedColor(ColorScheme.dark));
    }
  }
}

File: src/main/java/org/voegtle/weatherwidget/widget/SmallWidgetScreenPainter.kt
package org.voegtle.weatherwidget.widget

import android.appwidget.AppWidgetManager
import android.widget.RemoteViews
import org.voegtle.weatherwidget.R
import org.voegtle.weatherwidget.data.WeatherData
import org.voegtle.weatherwidget.preferences.ColorScheme
import org.voegtle.weatherwidget.util.ColorUtil

class SmallWidgetScreenPainter(appWidgetManager: AppWidgetManager, widgetIds: IntArray, private val remoteViews: RemoteViews) : AbstractWidgetScreenPainter(appWidgetManager, widgetIds, remoteViews) {

  override fun showDataIsInvalid() {
    remoteViews.setTextColor(R.id.weather_small, ColorUtil.updateColor(ColorScheme.dark))
  }

  override fun showDataIsValid() {
    updateAllWidgets()
  }

  fun updateData(data: WeatherData?) {
    if (data != null) {
      remoteViews.setTextViewText(R.id.weather_small, formatter.formatTemperature(data))
      remoteViews.setTextColor(R.id.weather_small, ColorUtil.byAge(data.timestamp!!))
    } else {
      remoteViews.setTextColor(R.id.weather_small, ColorUtil.outdatedColor(ColorScheme.dark))
    }
  }
}
